editor/src/clj/editor/pose.clj

```diff
@@ -0,0 +1,255 @@
+          (* vz qz qz)
+          (* vx qy qw -2.0)
+          (* vy qx qw 2.0)
+          (* vz qw qw))
+       (* vz qx qx)
+       (* vz qy qy))))
+
+(defn- rotate-quaternion [[^double ax ^double ay ^double az ^double aw] [^double bx ^double by ^double bz ^double bw]]
+  (vector-of
+    :double
+    (- (+ (* ax bw)
+          (* ay bz)
+          (* aw bx))
+       (* az by))
+    (- (+ (* ay bw)
+          (* az bx)
+          (* aw by))
+       (* ax bz))
+    (- (+ (* ax by)
+          (* az bw)
+          (* aw bz))
+       (* ay bx))
+    (- (* aw bw)
+       (* ax bx)
+       (* ay by)
+       (* az bz))))
+
+(defn translation-pose
+  ^Pose [^double x ^double y ^double z]
+  (if (and (zero? x) (zero? y) (zero? z))
+    default
+    (->Pose (vector-of :double x y z) default-rotation default-scale)))
+
+(defn rotation-pose
+  ^Pose [^double x ^double y ^double z ^double w]
+  (if (and (zero? x) (zero? y) (zero? z) (== 1.0 w))
+    default
+    (->Pose default-translation (vector-of :double x y z w) default-scale)))
+
+(defn euler-rotation [^double x-deg ^double y-deg ^double z-deg]
+  ;; Implementation based on:
+  ;; http://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/19770024290.pdf
+  ;; Rotation sequence: 231 (YZX)
+  (if (== 0.0 x-deg y-deg)
+    (if (== 0.0 z-deg)
+      default-rotation
+      (let [ha (* 0.5 (math/deg->rad z-deg))
+            s (Math/sin ha)
+            c (Math/cos ha)]
+        (vector-of :double 0.0 0.0 s c)))
+    (let [t1 (math/deg->rad y-deg)
+          t2 (math/deg->rad z-deg)
+          t3 (math/deg->rad x-deg)
+          c1 (Math/cos (* t1 0.5))
+          s1 (Math/sin (* t1 0.5))
+          c2 (Math/cos (* t2 0.5))
+          s2 (Math/sin (* t2 0.5))
+          c3 (Math/cos (* t3 0.5))
+          s3 (Math/sin (* t3 0.5))
+          c1c2 (* c1 c2)
+          s2s3 (* s2 s3)
+          qx (+ (* s1 s2 c3)
+                (* s3 c1c2))
+          qy (+ (* s1 c2 c3)
+                (* s2s3 c1))
+          qz (+ (- (* s1 s3 c2))
+                (* s2 c1 c3))
+          qw (+ (- (* s1 s2s3))
+                (* c1c2 c3))
+          dp (+ (* qx qx) (* qy qy) (* qz qz) (* qw qw))]
+      (if (> dp 0.0)
+        (let [r (/ 1.0 (Math/sqrt dp))]
+          (vector-of :double (* qx r) (* qy r) (* qz r) (* qw r)))
+        num4-zero))))
+
+(defn euler-rotation-pose
+  ^Pose [^double x-deg ^double y-deg ^double z-deg]
+  (if (and (zero? x-deg) (zero? y-deg) (zero? z-deg))
+    default
+    (->Pose default-translation (euler-rotation x-deg y-deg z-deg) default-scale)))
+
+(defn scale-pose
+  ^Pose [^double x ^double y ^double z]
+  (if (and (== 1.0 x) (== 1.0 y) (== 1.0 z))
+    default
+    (->Pose default-translation default-rotation (vector-of :double x y z))))
+
+(defn make
+  ^Pose [translation rotation scale]
+  {:pre [(or (nil? translation) (s/valid? ::translation translation))
+         (or (nil? rotation) (s/valid? ::rotation rotation))
+         (or (nil? scale) (s/valid? ::scale scale))]}
```

Comment:
I think we should create and prioritize an issue about disabling asserts in production, and going through our asserts to separate dev-time contracts like these from user input validation. The spec library is somewhat slow...

Reply:
Yeah, we probably should. Disabling asserts did not seem to make a lot of difference at the time I profiled it, but I've started to add more checks since then.

But I'm not really worried about this particular case, since we only create `Pose` instances when producing build targets.